File: ruby/total_spend.rb
require 'set'

class TotalSpend
  NOTABLE_CATEGORIES = Set["Insurance", "Copay"]

  def self.by_spend_type(invoice_items)
    total = {}

    invoice_items.each do |item|
      amount = total.fetch(item.spend_type, 0) + item.transaction.amount
      total[item.spend_type] = amount
    end

    total
  end

  def self.by_spend_type_and_category(invoice_items)
    total = {}

    invoice_items.each do |item|
      category = TotalSpend::NOTABLE_CATEGORIES.include?(item.transaction.category) ? item.transaction.category : "Other"
      prev_amount = total.dig(item.spend_type, category) || 0
      unless total.include?(item.spend_type)
        total[item.spend_type] = {}
      end
      total[item.spend_type][category] = prev_amount + item.amount
    end

    total
  end
end

File: python/total_spend.py
class TotalSpend:
  NOTABLE_CATEGORIES = {"Insurance", "Copay"}

  @staticmethod
  def by_spend_type(invoice_items):
    total = {}
    for item in invoice_items:
      amount = total.get(item.spend_type, 0) + item.transaction.amount
      total[item.spend_type] = amount
    return total

  @staticmethod
  def by_spend_type_and_category(invoice_items):
    total = {}
    for item in invoice_items:
      if item.transaction.category in TotalSpend.NOTABLE_CATEGORIES:
        category = item.transaction.category
      else:
        category = "Other"
      amount = total.get((item.spend_type, category), 0) + item.amount
      total[(item.spend_type, category)] = amount
    return total
